### PR TITLE
Support detecting cl.exe in vs2019

### PR DESCRIPTION
**Pull request by eternalphane**
Use `setuptools.msvc` instead of `distutils.msvc9compiler` to detect cl.exe

### CODE REVIEW

**Comment by kmaehashi** [comment]
Thanks @eternalphane!

It looks nice to discover VS2019, however I'm concerning that CUDA 10.0, which is still in our supported CUDA version, does not support VS2019.
https://docs.nvidia.com/cuda/archive/10.0/cuda-installation-guide-microsoft-windows/index.html
Let me try your patch on CUDA 10.0 environment later.

**Review by leofang — CHANGES_REQUESTED** [review]
Thanks for your patch. I do not have access to any Windows CUDA env, so I cannot verify the patch and I think @kmaehashi and others from the team will help. But, for the purpose of #4983 I think this PR is necessary for us to move away from depending `distutils` at runtime, so I am happy to see this change!

With the above background stated, @eternalphane could you try replacing all `distutils` occurrences in this file (likely only contained in the function you're touching) by their `setuptools` counterparts and help us verify the change please? I think PEP 632 should be clear in the needed replacements so it should be rather straightforward.

File: cupy/cuda/compiler.py [hunk 1]
```diff
@@ -89,21 +89,23 @@ def _run_cc(cmd, cwd, backend, log_stream=None):
 
 @_util.memoize()
 def _get_extra_path_for_msvc():
-    import distutils.spawn
-    cl_exe = distutils.spawn.find_executable('cl.exe')
+    import shutil
```
Comment (kmaehashi) [inline cupy/cuda/compiler.py:92]: Could you move this import (and `import platform` too --- as they're standard modules) to the top of the file?

File: cupy/cuda/compiler.py [hunk 2]
```diff
@@ -89,21 +89,23 @@ def _run_cc(cmd, cwd, backend, log_stream=None):
...
     if cl_exe:
         # The compiler is already on PATH, no extra path needed.
         return None
 
-    from distutils import msvc9compiler
-    vcvarsall_bat = msvc9compiler.find_vcvarsall(
-        msvc9compiler.get_build_version())
-    if not vcvarsall_bat:
+    from setuptools import msvc
+    from setuptools.errors import DistutilsError
```
Comment (kmaehashi) [inline cupy/cuda/compiler.py:99]: It seems `setuptools.errors` is only available in setuptools v42.0+. In my environment, the default one that comes with Python 3.8 for Windows was 41.2.0, so this line failed. Until we decide a policy on which setuptools versions to support (we should do this once we fully migrate from distutils), I'm negative to use this module. So, could you use `distutils.errors.DistutilsPlatformError` instead, as in your original code? :pray:

File: cupy/cuda/compiler.py [hunk 3]
```diff
@@ -89,21 +89,23 @@ def _run_cc(cmd, cwd, backend, log_stream=None):
...
+    try:
+        vctools = msvc.EnvironmentInfo(platform.machine().lower()).VCTools
+    except DistutilsError:
         # Failed to find VC.
         return None
 
-    path = os.path.join(os.path.dirname(vcvarsall_bat), 'bin')
-    if not distutils.spawn.find_executable('cl.exe', path):
+    path = vctools[-1]
+    if not shutil.which('cl.exe', path):
```
Comment (kmaehashi) [inline cupy/cuda/compiler.py:108]:
```suggestion
    if not shutil.which('cl.exe', path=path):
```

**Review by kmaehashi — CHANGES_REQUESTED** [review]

**Comment by kmaehashi** [comment]
Could you fix the comments here as well?

https://github.com/cupy/cupy/blob/b01641dedb065f65a4d3f846cf3c6222693a8e44/cupy/cuda/compiler.py#L54-L56

`the VC++ compiler used to build Python / CuPy to the PATH` -> `the latest VC++ compiler found via setuptools`

**Comment by kmaehashi** [comment]
I've tookover the work in #6025.

**Pull request closed without merging**